tests/upgrade_tests/test_firmware_upgrades.py

```diff
@@ -333,6 +337,41 @@ def test_upgrade_shamir_recovery(gen: str, tag: Optional[str]):
         assert state.mnemonic_type == BackupType.Slip39_Basic
 
 
+@for_all("core", core_minimum_version=(2, 1, 9))
+def test_upgrade_shamir_backup(gen: str, tag: Optional[str]):
+    with EmulatorWrapper(gen, tag) as emu:
+        device.reset(
+            emu.client,
+            pin_protection=False,
+            skip_backup=True,
+            backup_type=BackupType.Slip39_Basic,
+        )
+        device_id = emu.client.features.device_id
+        backup_type = emu.client.features.backup_type
+        mnemonic_secret = emu.client.debug.state().mnemonic_secret
+        address = btc.get_address(emu.client, "Bitcoin", PATH)
+        assert emu.client.features.needs_backup
+        storage = emu.get_storage()
+
+    with EmulatorWrapper(gen, storage=storage) as emu:
+        assert emu.client.features.device_id == device_id
+        assert emu.client.features.backup_type == backup_type
+        assert btc.get_address(emu.client, "Bitcoin", PATH) == address
+        assert emu.client.features.needs_backup
+
+        with emu.client:
+            IF = InputFlowSlip39BasicBackup(emu.client, False)
+            emu.client.set_input_flow(IF.get())
+            device.backup(emu.client)
+
+        assert not emu.client.features.needs_backup
+
+        # Check result
+        groups = shamir.decode_mnemonics(IF.mnemonics[:3])
+        ems = shamir.recover_ems(groups)
+        assert ems.ciphertext == mnemonic_secret
```

Comment:
I suspect that testing this is not enough. We (also?) need to test that passphrased addresses come out the same after upgrade.

I would maybe also add a version-gated check that if you set up your device on fw \< 2.7.1, the backup will _not_ have the extendable flag despite being generated on 2.7.1 and up

Reply:
> I would maybe also add a version-gated check that if you set up your device on fw < 2.7.1, the backup will not have the extendable flag despite being generated on 2.7.1 and up

https://github.com/trezor/trezor-firmware/pull/3825/commits/0b480e7d3b6799644e371ddcb318207bb872c248

> I suspect that testing this is not enough. We (also?) need to test that passphrased addresses come out the same after upgrade.

I do check on [line 358](https://github.com/trezor/trezor-firmware/pull/3825/files#diff-08c67d05fcaafcc8efd599cf286f0aeda1f219e5b802f0dcae847c54416cb1d6R358) that an address of the standard wallet remains the same, which should be enough, considering that a standard wallet is a passphrased wallet with empty passphrase.

```py
assert btc.get_address(emu.client, "Bitcoin", PATH) == address
```

I am having trouble implementing the same check for a passphrased wallet. The following code returns the same address, even if I set `passphrase_protection=True` in `ResetDevice` and `_passphrase_source=2` in `ApplySettings`. Not sure whether it's worth figuring out.

```py
emu.client.use_passphrase("TREZOR")
btc.get_address(emu.client, "Bitcoin", PATH)
```

Reply:
I believe you'd need to insert an `emu.client.initialize(new_session=True)` before switching passphrase.

you are _probably_ right that the standard wallet address is a good check, but using two different passphrases is probably a better check either way

Reply:
https://github.com/trezor/trezor-firmware/pull/3825/commits/951ac292f3af662974201d69fc6f5a2278661ad1